File: Python/Rotate Image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return
        
        # rotate nums at each level
        n = len(matrix) - 1
        current = int(n / 2)
        while current >= 0:
            for index in range(n - current * 2):
                # nums on four sides
                matrix[current][current + index], matrix[current + index][n - current], matrix[n - current][n - current - index], matrix[n - current - index][current] = matrix[n - current - index][current], matrix[current][current + index], matrix[current + index][n - current], matrix[n - current][n - current - index]
            current -= 1

File: C++/Rotate Image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        if(matrix.size() < 2)
            return;
        
        int n = matrix.size() - 1;
        int iteration = (int) (n / 2);
        
        // iteration from 0,0 -> center n/2,n/2
        for(int i = 0; i <= iteration; i++){
            for(int j = 0; j < n - 2 * i; j++){
                cout << i << j << endl;
                int temp = matrix[i][i + j];
                matrix[i][i + j] = matrix[n - i - j][i];
                matrix[n - i - j][i] = matrix[n - i][n - i - j];
                matrix[n - i][n - i - j] = matrix[i + j][n - i];
                matrix[i + j][n - i] = temp;
            }
        }
    }
};